README.md
# sveltekit-medusa-client

A client library for communicating with a Medusa ecommerce backend in SvelteKit

[Documentation](https://pevey.com/sveltekit-medusa-client)

If you are not familiar with Medusa, you can learn more on [the project web site](https://www.medusajs.com/).

> Medusa is a set of commerce modules and tools that allow you to build rich, reliable, and performant commerce applications without reinventing core commerce logic. The modules can be customized and used to build advanced ecommerce stores, marketplaces, or any product that needs foundational commerce primitives. All modules are open-source and freely available on npm.

This client is designed to be used on the server.  It cannot be exported to the browser.  This means you must make your calls to your Medusa backend from your storefront server, not from the client browser.  Calls to the library can be made from:

* A handler in `hooks.server.js/ts`
* A page load function in `+page.server.js/ts`
* A form action in `+page.server.js/ts`, or
* An API endpoint, aka `+server.js/ts`

One of the benefits of newer frameworks like SvelteKit is that they combine the fluid user experience of client-side reactivity with the ability to handle logic on the server when you choose to.  Keeping your Medusa backend firewalled and accessible only to your storefront application server provides an additional layer of security versus having your backend directly exposed.  This type of deployment also allows us to use tools like Turnstile or reCAPTCHA to provide some protection against bots and brute force attacks.  Without firewalling your backend, it would not be of much use to implement turnstile protection on your frontend.  It could easily be bypassed.

## Example Project

You can view an example project using this client library [here](https://github.com/pevey/sveltekit-medusa-starter).

## Installation

Create a new SvelteKit app if needed.  Then, install this package.

```bash

yarn add sveltekit-medusa-client

```

You should set the location of your Medusa server as an environment variable.  For example:

`.env`

```bash
MEDUSA_BACKEND_URL="http://localhost:9000"
```

## Basic Usage

To create a new client, invoke the MedusaClient constructor, passsing the location of your Medusa server as an argument.  For example:

`+page.server.js`

```js
import { MedusaClient } from 'sveltekit-medusa-client'
import { MEDUSA_BACKEND_URL } from '$env/static/private'

export const load = async function () {
   const medusa = new MedusaClient(MEDUSA_BACKEND_URL)
   return {
      products: medusa.getProducts()
   }
}
```

Then, on the corresponding `+page.svelte`, you can use the products data you exported:
(For more information on the data returned, refer to the [Medusa API Documentation](https://docs.medusajs.com/api/store#tag/Products/operation/GetProducts))

```svelte
<script>
   export let data
   const products = data.products  
</script>

<ul>
{#each products as product}
   <li>
      Product id: {product.id}<br>
      Product handle: {product.handle}<br>
      {product.title}
   </li>
{:else}
   <p>No products returned</p>
{/each}
<ul>
```

## Using the Client as a Singleton

One major drawback of the example above is that a new Medusa client is created for each page load.  
You can prevent that by adding a small library in your project that creates a single shared client that can be imported where needed.
For example:

`lib/server/medusa.js`

```js
import { MedusaClient } from 'sveltekit-medusa-client'
import { MEDUSA_BACKEND_URL } from '$env/static/private'
export default new MedusaClient(MEDUSA_BACKEND_URL)
```

Now, on our `+page.server.js` load function, we can do this:

```js
import medusa from '$lib/server/medusa'

export const load = async function () {
   return {
      products: medusa.getProducts()
   }
}
```

## Client Options

A number of options give some flexibility to the client.  The options object that can be injected in the client contructor takes this shape:

```ts
export interface ClientOptions {
   retry?: number
   timeout?: number
   headers?: {}
   persistentCart?: boolean
   logger?: Logger
   logFormat?: 'text' | 'json' | 'majel'
   logLevel?: 'verbose' | 'limited' | 'silent'
   excludedPaths?: string[]
   limitedPaths?: string[]
}
```

For example, you can create a new client instance like this:

```js
import { MedusaClient } from 'sveltekit-medusa-client'
import { MEDUSA_BACKEND_URL, CLOUDFLARE_ACCESS_ID, CLOUDFLARE_ACCESS_SECRET } from '$env/static/private'
export default new MedusaClient(MEDUSA_BACKEND_URL, { 
   timeout: 3000, // 3 seconds
   retry: 0,
   headers: {
      'CF-Access-Client-Id': CLOUDFLARE_ACCESS_ID,
      'CF-Access-Client-Secret': CLOUDFLARE_ACCESS_SECRET,
   },
   persistentCart: true,
   logger: console,
   logFormat: 'json',
   logLevel: 'verbose',
   excludedPaths: ['/store/mycustomsensitiveroute'],
   limitedPaths: ['/store/bulkyresponseroute']
})
```

- `timeout` - The default is 8000, or 8 seconds.  The length of time to wait for a response before aborting 
- `retry` - The default is 3.  The number of times to retry a timed out request
- `headers` - The default is undefined.  An object of HTTP headers, as many as you want, which will be added to all requests sent to the backend.  This can be useful in many situations.  If you would like to access a server behind a proxy with bearer auth, you can pass the auth header in this property.  You can also pass Cloudflare Access service auth credentials, as in the example above.
- `persistentCart` - The default is false.  If true, the client will expect an endpoint at `/store/customers/me/cart` that will return the customer's cart.  For now, this endpoint is not included in the Medusa core and must be added.
- `logger` - The default is `console`.  You can inject your own logger instance if you already have one configured in the application.  For example, a winston logger instance.  Any logger that implements the `info()` and `error()` methods should work.
- `logFormat` - The default is json.  You can change to 'text' if you need to for some reason.
- `excludedPaths` - The default is ['/store/auth'].  An array of strings that should be checked to exclude paths from logging.  The default can be added to, but not overridden.  Requests to URIs on your medusa backend that contain one or more of these strings will not be logged.  
- `limitedPaths` - The default is undefined.  An array of strings that should be checked to reduce the level of detail when logging.  Requests to URIs on your medusa backend that contain one or more of these strings will not log request or response content, only metadata.  The url of the request will be logged, but not query params.

## Authentication

Some methods in the library, like the `getProducts` method in the example above, need no authentication.  Other methods need more context, such as whether the requester is a logged in user, or whether they have an existing shopping cart.  The first argument passed to those methods is the special SvelteKit `locals` object.  Locals on the server work much like a page or session store in the browser.  They are a place to hold on to data related to this particular request that we may need somewhere else in the application before this request/response cycle is complete.

Use the middleware method `handleRequest` from this library to handle customer authentication on every request with very little effort.  If the user is logged in, the user object will be available at `locals.user.` Middleware is added in SvelteKit via the hooks.server.js/ts file:

`hooks.server.js`

```js
import medusa from '$lib/server/medusa'

export const handle = async ({ event, resolve }) => {
   event = await medusa.handleRequest(event)
   return await resolve(event)
}
```

Now, we can invoke methods that require information about the user and the cart.

`+page.server.js`

```js
import medusa from '$lib/server/medusa'

export const load = async function ({ locals, cookies }) {
   return {
      cart: medusa.getCart(locals, cookies)
   }
}
```

## Caching

Caching is enabled by passing a key string in the options for the functions that support caching. The key is the unique identifier for that particular query response.  Optionally, you can also pass a ttl. The ttl is the max age of the cache in milliseconds.  The default ttl is 1000.  

### Caching Example

To enable caching on a call to the getProduct function (`medusa.getProduct(handle)`), call the function like this:

```js
let product = await medusa.getProduct(handle, { key: `__${handle}__product`, ttl: 10000 })
```

Behind the scenes, the response will be cached in memory for the duration of the ttl.

### Important Notes

- Short cache times are recommended.  Even a short ttl can lead to a significant performance boost in your storefront application and reduction of load on your Medusa backend on high traffic sites.

- The cache is stored in memory.  This is ideal in some scenarios, but not in memory-constrained environments or for especially large sites.

- When deploying to a serverless platform, you will probably want to use something like Redis in your storefront application for caching and forgo the built-in cache option.


CHANGELOG.md
# Change Log

## 3.1.2

### Patch Changes

- Removed client option 'disableCache' from interface definition to avoid confuction since it is no longer used.  Cache is not enabled by default and is instead enabled when a function is called for that specific function call.

## 3.1.1

### Patch Changes

- Added cache option: revalidate  If a cache key is passed and revalidate is set to true, this will cause the query to pull fresh data, cache the new data, and update the ttl.

## 3.1.0

### Patch Changes

- Caching API change:  Caching is no longer set at the client/class level.  Caching is now enabled by passing a key string in the options for particular functions that support caching. The key is the unique identifier for that particular query response.  Optionally, you can also pass a ttl. The ttl is the max age of the cache in milliseconds.  The default ttl is 1000.  See the latest README for more information.

The list of functions that support caching:

```js
getSearchResults(q:string, cacheOptions?:CacheOptions)
getProducts(options?:ProductRetrievalOptions, cacheOptions?:CacheOptions)
getCollections(options?:CollectionRetrievalOptions, cacheOptions?:CacheOptions)
getCollection(handle:string, cacheOptions?:CacheOptions)
getCollectionProducts(id:string, options?:ProductRetrievalOptions, cacheOptions?:CacheOptions)
getProduct(handle:string, cacheOptions?:CacheOptions)
getReviews(productId:string, options?:ReviewRetrievalOptions, cacheOptions?:CacheOptions)
```

Example: To enable caching for the getProduct function (`medusa.getProduct(handle)`), call the function like this:

```js
let product = await medusa.getProduct(handle, { key: `${params.slug}_product`, ttl: 10000 })
```

## 3.0.0

### Patch Changes

- Fix: No more ts warning about type mismatch when not using custom headers (by @ellicodan)
- Chore: Update sveltekit-superfetch dependency to ^3.0
- Feat: Enabled caching of product, collection, and search requests by default.  To disable, in options object set 'disableCache' to true.

## 2.0.2

### Patch Changes

- Fix: Bug in 2.0 that caused the login function to return 'true' even though the auth failed and the user was not logged in (session cookie not set).  This bug did not affect access but could cause UX issues in the login flow.

## 2.0.0

### Patch Changes

- Chore: Bumped sveltekit-superfetch dependency to ^2.0.  This changed the syntax slightly for sending queries, which you should only notice if you were using the public query() method directly.
- Feat: Enabled optional logging.  You can inject a winston logger or other logger instance into the MedusaClient constructor.  If 'debug' is set to true but no logger instance is passed, console will be used by default.  See updated docs.

## 1.13.2

### Patch Changes

- Fix: Update User interface to make phone property optional

## 1.13.0

### Patch Changes

- Feat: Added typing for product-related functions.  Will add typing for Users, Carts, and other return values as Medusa team moves them over to @medusajs/types package.
- Fix: Bug where getCart function could return a cart that had been completed on a different device if persistentCart was set to true in certain situations.
- Chore: Versioning changed.  Version numbers will now follow semantic versioning and no longer mirror medusa backend versioning.

## 1.12.1-c

### Patch Changes

- Chore: Compatibility with latest medusa-plugin-ratings v1.3

## 1.12.1-b

### Patch Changes

- Chore: Bump versions on all dependencies to latest

## 1.12.1

### Patch Changes

- Fix: Update method of parsing set-cookie headers to account for other headers being set other than the Medusa connect.sid session header

## 1.12.0-b

### Patch Changes

- Fix: getCustomer function: Move fetch request to get customer session back inside try block to enable smoother error handling.  It was left outside of the block by accident after debugging.
- Fix: Bug with new persistent cart feature that caused an error when user logged in and did not already have any cart sessions
- Fix: Bug in the first login after registering a user that caused register to return false even though it was successful

## 1.12.0

### Patch Changes

- Feat: Session cookie expiration now matches any custom ttl set in medusa.config.js
- Feat: Session cookie now supports rolling:true (refresh) session option in medusa.config.js
- Feat: Options object passed to constructor now supports custom timeout (in milliseconds) and retry settings.

src/lib/index.ts
import cookie from 'cookie'
import { SuperFetch } from 'sveltekit-superfetch'
import type { ProductDTO } from '@medusajs/types'
import type { Cookies, RequestEvent } from '@sveltejs/kit'
import { dev } from '$app/environment'

export interface CacheOptions {
   key: string
   ttl?: number
}

export interface ProductRetrievalOptions {
   limit?: number
   offset?: number
   order?: string
   expand?: string
   fields?: string
   query?: string
}

export interface CollectionRetrievalOptions {
   limit?: number
   offset?: number
}

export interface ReviewRetrievalOptions {

}

export interface Review {
   id?: string
   product_id: string
   customer_id?: string
   display_name: string
   content: string
   rating: number
   approved?: boolean
}

export interface User {
   // used in register()
   // should match this request schema:
   // https://docs.medusajs.com/api/store#tag/Customers/operation/PostCustomers
   first_name: string
   last_name: string
   email: string
   password: string
   phone?: string
}

export interface Customer {
   // used in editCustomer()
   // should match this request schema:
   // https://docs.medusajs.com/api/store#tag/Customers/operation/PostCustomersCustomer
   first_name?: string
   last_name?: string
   billing_address?: Address
   password?: string
   phone?: string
   email?: string
   metadata?: object
}

export interface Address {
   // used in Customer interface, updateCartAddress(), addAddress(), and updateShippingAddress()
   // should match this request schema:
   // https://docs.medusajs.com/api/store#tag/Customers/operation/PostCustomersCustomerAddresses
   first_name: string
   last_name: string
   phone?: string
   company?: string
   address_1: string
   address_2?: string
   city: string
   country_code: string
   province: string
   postal_code: string
   metadata?: object
}

export interface ClientOptions {
   retry?: number
   timeout?: number
   headers?: {}
   persistentCart?: boolean
   logger?: Logger
   logFormat?: 'text' | 'json' | 'majel'
   logLevel?: 'verbose' | 'limited' | 'silent'
   excludedPaths?: string[]
   limitedPaths?: string[]
}

export interface QueryOptions {
   locals?: App.Locals
   path: string
   method?: string
   body?: object
   key?: string
   ttl?: number
   revalidate?: boolean
   logLevel?: 'verbose' | 'limited' | 'silent'
}

export interface Logger {
   info: (message: string) => void
   error: (message: string) => void
}

export class MedusaClient {
   private url: string
   private timeout: number = 8000
   private retry: number = 0
   private headers: any
   private persistentCart: boolean = false
   private logger: Logger = console
   private logFormat: 'text' | 'json' | 'majel' = 'json'
   private logLevel: 'verbose' | 'limited' | 'silent' = (dev)? 'limited' : 'silent'
   private excludedPaths: string[] = ['/store/auth']
   private limitedPaths: string[] = []
   private superFetch: SuperFetch

   constructor(url: string, options?: ClientOptions) {
      this.url = url
      if (options) {
         let { timeout, retry, headers, persistentCart, logger, logFormat, logLevel, excludedPaths, limitedPaths } = options
         if (timeout) this.timeout = timeout
         if (retry) this.retry = retry
         if (headers) this.headers = headers
         if (persistentCart) this.persistentCart = persistentCart
         if (logger) this.logger = logger
         if (logFormat) this.logFormat = logFormat
         if (logLevel) this.logLevel = logLevel
         if (excludedPaths) { 
            for (const excluded of excludedPaths) {
               this.excludedPaths.push(excluded) 
            }
         }
         if (limitedPaths) {
            for (const limited of limitedPaths) {
               this.limitedPaths.push(limited)
            }
         }
      }
      this.superFetch = new SuperFetch({
         retry: this.retry,
         timeout: this.timeout,
         logger: this.logger,
         logFormat: this.logFormat,
         logLevel: this.logLevel,
         excludedPaths: this.excludedPaths,
         limitedPaths: this.limitedPaths
      })
   }

   async query(options: QueryOptions): Promise<Response|null> {
      const { locals, path, method = 'GET', body = {}, ...rest } = options
      let headers: any = {}
      if (this.headers) {
        for (const [key, value] of Object.entries(this.headers)) {
           headers[key] = value
        }
      }
      if (locals && locals.sid) {
         headers['Cookie'] = `connect.sid=${locals.sid}`
      }
      if (Object.keys(body).length != 0) {
         headers['Content-Type'] = 'application/json'
      }
      return await this.superFetch.query({
         url: `${this.url}${path}`,
         method,
         headers,
         body: (Object.keys(body).length != 0) ? JSON.stringify(body) : null,
         ...rest
      }).catch((e: Error) => {
         console.log(e)
         return null
      })
   }

   buildQuery(base: string, options:any = {}) {
      let queryString = base
      if (Object.keys(options).length !== 0) queryString += '?'
      if (options.limit) queryString += `limit=${options.limit}&`
      if (options.offset) queryString += `offset=${options.offset}&`
      if (options.order) queryString += `order=${options.order}&`
      if (options.expand) queryString += `expand=${encodeURIComponent(options.expand)}&`
      if (options.fields) queryString += `fields=${encodeURIComponent(options.fields)}&`
      if (options.query) queryString += `${encodeURIComponent(options.query)}&`
      return queryString
   }

   async handleRequest(event:RequestEvent) {
      // this middleware function is called by src/hooks.server.ts or src/hooks.server.js

      event.locals.sid = event.cookies.get('sid')
      if (event.locals.sid) event.locals.user = await this.getCustomer(event.locals, event.cookies) 
      else event.locals.sid = ''

      event.locals.cartid = event.cookies.get('cartid')
      let cart: any = await this.getCart(event.locals, event.cookies)
      event.locals.cartid = cart?.id || ''
      event.locals.cart = cart || null

      return event
   }

   async parseAuthCookie(setCookie:[] = [], locals:App.Locals , cookies:Cookies) {
      if (!setCookie) return false
      try {
         for (let rawCookie of setCookie) {
            let parsedCookie = cookie.parse(rawCookie)
            if (parsedCookie['connect.sid']) {
               locals.sid = parsedCookie['connect.sid']
               let expires = new Date(parsedCookie['Expires'])
               let maxAge = Math.floor((expires.getTime() - Date.now()) / 1000)
               cookies.set('sid', locals.sid, {
                  path: '/',
                  maxAge: maxAge,
                  sameSite: 'strict',
                  httpOnly: true,
                  secure: true
               })
               return true
            }
         }
      } catch (e) {
         console.log(e)
         return false
      }
   }

   async getCustomer(locals:App.Locals, cookies:Cookies) {
      // returns a user object if found, or null if not
      return await this.query({ 
         locals, 
         path: '/store/auth' 
      }).then((response:any) => {
         this.parseAuthCookie(response.headers.getSetCookie(), locals, cookies)
         return response.json().then((data:any) => data.customer)
      }).catch(() => null)
   }

   async login(locals:App.Locals, cookies:Cookies, email:string, password:string) {
      // returns true or false based on success
      const response = await this.query({
         locals, 
         path: '/store/auth', 
         method: 'POST', 
         body: { email, password },
         logLevel: 'silent'
      })
      if (!response || !response.ok) return false
      // @ts-ignore, getSetCookie() is new and not yet in the type definition for Headers, but it is valid
      return await this.parseAuthCookie(response.headers?.getSetCookie(), locals, cookies).catch(() => false)
   }

   async logout(locals:App.Locals, cookies:Cookies) {
      // returns true or false based on success
      let success = await this.query({
         locals, 
         path: '/store/auth', 
         method: 'DELETE'
      }).then((res:any) => res.ok).catch(() => false)
      if (!success) return false
      locals.sid = ''
      locals.user = {}
      cookies.delete('sid')
      return true
   }

   async register (locals:App.Locals, cookies:Cookies, user:User) {
      // returns true or false based on success
      const { email, password } = user
      return await this.query({
         locals, 
         path: '/store/customers', 
         method: 'POST', 
         body: user,
         logLevel: 'silent'
      }).then((res:any) => {
         if (res.ok) { 
            return this.login(locals, cookies, email, password).then(() => true).catch(() => false) 
         } else return false
      }).catch(() => false)
   }

   async getSearchResults(q:string, cacheOptions?:CacheOptions) {
      // returns an array of hits, if any
      if (!q) { return Array() }
      return await this.query({
         path: '/store/products/search',
         method: 'POST',
         body: { q },
         ...cacheOptions
      }).then((res:any) => res.json()).then((data:any) => data.hits).catch(() => null)
   }

   async getProducts(options?:ProductRetrievalOptions, cacheOptions?:CacheOptions): Promise<ProductDTO[]|null> {
      // returns an array of product objects
      const queryString = this.buildQuery('/store/products', options)
      return await this.query({ path: queryString, ...cacheOptions })
      .then((res:any) => res.json()).then((data:any) => data.products).catch(() => null)
   }

   async getCollections(options?:CollectionRetrievalOptions, cacheOptions?:CacheOptions) {
      // returns an array of collection objects on success
      const queryString = this.buildQuery('/store/collections', options)
      return await this.query({ path: queryString, ...cacheOptions })
      .then((res:any) => res.json()).then((data:any) => data.collections).catch(() => null)
   }

   async getCollection(handle:string, cacheOptions?:CacheOptions) {
      // returns a collection object on success
      return await this.query({ path: `/store/collections?handle[]=${handle}`, ...cacheOptions })
      .then((res:any) => res.json()).then((data:any) => data.collections[0]).catch(() => null)
   }

   async getCollectionProducts(id:string, options?:ProductRetrievalOptions, cacheOptions?:CacheOptions): Promise<ProductDTO[]|null> {
      // returns an array of product objects on success
      let base = `/store/products?collection_id[]=${id}`
      const queryString = this.buildQuery(base, options)
      return await this.query({ path: queryString, ...cacheOptions })
      .then((res:any) => res.json()).then((data:any) => data.products).catch(() => null)
   }

   async getProduct(handle:string, cacheOptions?:CacheOptions): Promise<ProductDTO|null> {
      // returns a product object on success
      let product = await this.query({ path: `/store/products?handle=${handle}`, ...cacheOptions })
      .then((res:any) => res.json()).then((data:any) => data.products[0]).catch(() => null)
      if (!product) { return null }
      for (let option of product.options) {
         option.filteredValues = this.filteredValues(option)
      }
      return product
   }

   async getReviews(productId:string, options?:ReviewRetrievalOptions, cacheOptions?:CacheOptions) {
      // returns an array of review objects on success
      // options - page = 1, limit = 10, sort = 'created_at', order = 'desc', search = null
      // TODO: handle options
      return await this.query({ path: `/store/products/${productId}/reviews`, ...cacheOptions })
      .then((res:any) => res.json()).then((data:any) => data.product_reviews).catch(() => null)
   }

   async getCustomerReviews(locals:App.Locals, options?:ReviewRetrievalOptions) {
      // returns an array of review objects on success
      // options - page = 1, limit = 10, sort = 'created_at', order = 'desc', search = null
      // TODO: handle options
      return await this.query({ locals, path: `/store/customers/me/reviews` })
      .then((res:any) => res.json()).then((data:any) => data.product_reviews).catch(() => null)
   }

   async getReview(reviewId:string) {
      // returns a review object on success
      return await this.query({ path:`/store/reviews/${reviewId}` })
      .then((res:any) => res.json()).then((data:any) => data.product_review).catch(() => null)
   }

   // CHANGE TO RETURN THE REVIEW OBJECT ON SUCCESS
   async addReview(locals:App.Locals, review:Review) {
      // returns true or false based on success
      return await this.query( {
         locals, 
         path: `/store/products/${review.product_id}/reviews`, 
         method: 'POST', 
         body: review
      })
      .then((res:any) => res.ok).catch(() => false)
   }

   async updateReview(locals:App.Locals, reviewId:string, review:Review) {
      // returns a review object on success, or null on failure
      return await this.query({ 
         locals, 
         path: `/store/reviews/${reviewId}`, 
         method: 'POST', 
         body: review
      }).then((res:any) => res.ok).catch(() => null)
   }

   async getCart(locals:App.Locals, cookies:Cookies) {
      // returns a cart array on success, otherwise null
      let cart
      if (locals.cartid) {
         cart = await this.query({ locals, path: `/store/carts/${locals.cartid}` })
         .then((res:any) => res.json()).then((data:any) => data.cart).catch(() => null)
         // if this cart was completed on another device, we don't want to use it
         if (cart && cart.completed_at) cart = null
      } else if (this.persistentCart && locals.user) {
         cart = await this.query({ locals, path: `/store/customers/me/cart` })
         .then((res:any) => res.json()).then((data:any) => data.cart).catch(() => null)
         if (cart) {
            cookies.set('cartid', cart.id, {
               path: '/',
               maxAge: 60 * 60 * 24 * 400,
               sameSite: 'strict',
               httpOnly: true,
               secure: true
            })
         }
      }
      if (locals.cartid && !cart) {
         locals.cartid = ''
         cookies.delete('cartid')
      }
      return cart
   }

   async addToCart(locals:App.Locals, cookies:Cookies, variantId:string, quantity:number = 1) {
      // returns a cart array on success, otherwise null
      if (!variantId) { return null }

      // try adding to existing cart
      if (locals.cartid) { 
         try {
            const cart = await this.query({
               locals, 
               path: `/store/carts/${locals.cartid}/line-items`, 
               method: 'POST', 
               body: { variant_id: variantId, quantity: quantity }
            }).then((res:any) => res.json()).then((data:any) => data.cart)
            return cart
         } catch {}
      }

      // if no cart or add to cart fails, try to create new cart
      const cart = await this.query({
         locals, 
         path: '/store/carts', 
         method: 'POST', 
         body: { items: [{ variant_id: variantId, quantity: quantity }] }
      }).then((res:any) => res.json()).then((data:any) => data.cart).catch(() => null )
      cookies.set('cartid', cart.id, {
         path: '/',
         maxAge: 60 * 60 * 24 * 400,
         sameSite: 'strict',
         httpOnly: true,
         secure: true
      })
      locals.cartid = cart.id

      return cart
   }

   async removeFromCart(locals:App.Locals, itemId:string) {
      // returns a cart array on success, otherwise null
      if (!locals.cartid || !itemId) { return null }
      return await this.query({
         locals, 
         path: `/store/carts/${locals.cartid}/line-items/${itemId}`, 
         method: 'DELETE'
      })
      .then((res:any) => res.json()).then((data:any) => data.cart).catch(() => null)
   }

   async updateCart(locals:App.Locals, itemId:string, quantity:number) {
      // returns a cart array on success, otherwise null
      if (!locals.cartid || !itemId || !quantity) { return null }
      return await this.query({
         locals, 
         path: `/store/carts/${locals.cartid}/line-items/${itemId}`, 
         method: 'POST', 
         body: { quantity: quantity }
      }).then((res:any) => res.json()).then((data:any) => data.cart).catch(() => null)
   }

   async updateCartBillingAddress(locals:App.Locals, address:Address) {
      // returns a cart array on success, otherwise null
      if (!locals.cartid) { return null }
      return await this.query({
         locals, 
         path: `/store/carts/${locals.cartid}`, 
         method: 'POST', 
         body: { billing_address: address }
      }).then((res:any) => res.json()).then((data:any) => data.cart).catch(() => null)
   }

   async updateCartShippingAddress(locals:App.Locals, address:Address) {
      // returns a cart array on success, otherwise null
      if (!locals.cartid) { return null }
      return await this.query({
         locals, 
         path: `/store/carts/${locals.cartid}`, 
         method: 'POST', 
         body: { shipping_address: address }
      }).then((res:any) => res.json()).then((data:any) => data.cart).catch(() => null)
   }

   async getShippingOptions(locals:App.Locals) {
      // returns an array of shipping option objects on success, otherwise null
      if (!locals.cartid) { return false }
      return await this.query({ locals, path: `/store/shipping-options/${locals.cartid}` })
      .then((res:any) => res.json()).then((data:any) => data.shipping_options).catch(() => null)
   }

   async selectShippingOption(locals:App.Locals, shippingOptionId:string) {
      // returns a cart array on success, otherwise null
      if (!locals.cartid || !shippingOptionId) { return null }
      return await this.query({
         locals, 
         path: `/store/carts/${locals.cartid}/shipping-methods`, 
         method: 'POST', 
         body: { option_id: shippingOptionId }
      }).then((res:any) => res.json()).then((data:any) => data.cart).catch(() => null)
   }

   async createPaymentSessions(locals:App.Locals) {
      // returns a cart array on success, otherwise null
      if (!locals.cartid) { return null }
      return await this.query({
         locals, 
         path: `/store/carts/${locals.cartid}/payment-sessions`, 
         method: 'POST'
      }).then((res:any) => res.json()).then((data:any) => data.cart).catch(() => null)
   }

   async selectPaymentSession(locals:App.Locals, providerId:string) {
      // returns a cart array on success, otherwise null
      if (!locals.cartid) { return null }
      return await this.query({
         locals, 
         path: `/store/carts/${locals.cartid}/payment-session`, 
         method: 'POST', 
         body: { provider_id: providerId }
      }).then((res:any) => res.json()).then((data:any) => data.cart).catch(() => null)
   }

   async completeCart(locals:App.Locals) {
      // returns an order object on success, otherwise null
      if (!locals.cartid) { return null }
      const reply = await this.query({
         locals, 
         path: `/store/carts/${locals.cartid}/complete`, 
         method: 'POST'
      }).then((res:any) => res.json()).catch(() => null )
      return (reply.type === 'order') ? reply.data : false
   }

   async addShippingAddress(locals:App.Locals, address:Address) {
      // returns true or false based on success
      if (!locals.user) { return false }
      return await this.query({
         locals, 
         path: `/store/customers/me/addresses`, 
         method: 'POST', 
         body: { address }
      }).then((res:any) =>  res.ok ).catch(() => false)
   }

   async updateShippingAddress(locals:App.Locals, addressId:string, address:Address) {
      // returns true or false based on success
      if (!locals.user) { return false }
      return await this.query({
         locals, 
         path: `/store/customers/me/addresses/${addressId}`, 
         method: 'POST', 
         body: address
      }).then((res:any) => res.ok).catch(() => false)
   }

   async deleteAddress(locals:App.Locals, addressId:string) {
      // returns true or false based on success
      if (!locals.user) { return false }
      return await this.query({
         locals, 
         path: `/store/customers/me/addresses/${addressId}`, 
         method: 'DELETE'
      }).then((res:any) => res.ok).catch(() => false)
   }

   async getAddresses(locals:App.Locals) {
      // returns an array of address objects on success, otherwise null
      if (!locals.user) { return null }
      return await this.query({ locals, path: `/store/customers/me/addresses` })
      .then((res:any) => res.json()).then((data:any) => data.addresses).catch(() => null)
   }

   async getOrder(locals:App.Locals, id:string) {
      // returns an order object on success, otherwise null
      return await this.query({ locals, path: `/store/orders/${id}` })
      .then((res:any) => res.json()).then((data:any) => data.order).catch(() => null)
   }

   async editCustomer(locals:App.Locals, customer:Customer) {
      // returns true or false based on success
      if (!locals.user) { return false }
      return await this.query({
         locals, 
         path: '/store/customers/me', 
         method: 'POST', 
         body: customer
      }).then((res:any) => res.ok).catch(() => false)
   }

   async requestResetPassword(email:string) {
      // returns true or false based on success
      return await this.query({ 
         path: '/store/customers/password-token', 
         method: 'POST', 
         body: { email } 
      }).then((res:any) => res.ok).catch(() => false)
   }

   async resetPassword(email:string, password:string, token:string) {
      // returns true or false based on success
      return await this.query({ 
         path: '/store/customers/password-reset', 
         method: 'POST', 
         body: { email, password, token }
      }).then((res:any) => res.ok).catch(() => false)
   }

   // @ts-ignore
   onlyUnique = (value, index, self) => self.indexOf(value) === index

   // @ts-ignore
   filteredValues = (option) => option.values.map((v) => v.value).filter(this.onlyUnique)
}